### PR TITLE
fix customElement.get

### LINKED ISSUE #3133: jsdom 16.5.0 break support for customElements.get
### Basic info:

- Node.js version: v14.13.0 
- jsdom version: 16.5.0 

Issue has been reproduce using last version of jest.

### Minimal reproduction case

(minimal setup using `npm init -y` in an empty directory and just adding jsdom as dependency in the project)

```js
const {JSDOM} = require('jdsom')

const dom = new JSDOM(`<body>
<script>
class CustomElement extends HTMLElement {}
customElements.define('my-ce', CustomElement )
console.log(customElements.get('my-ce'))
</script>
</body>`, {runScripts: 'dangerously'})
```

with 16.4.0 ==> works as expected
```
$ node index.js
[class CustomElement extends HTMLElement]
```

with 16.5.0 ==>  Does not work as expected
```
$ node index.js
undefined
```

---

The origin come from this refactoring : https://github.com/jsdom/jsdom/commit/ff69a75706eac5ef90d9eb46aedd90a14e29984c

In the file [CustomElementRegistry.html line 184](https://github.com/jsdom/jsdom/blame/2d82763248296c5f118d6fd890317f5db5de1635/lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js#184) the property `ctor` is removed :

```js
    const definition = {
      name,
      localName,
      // removed line  ctor,
      constructor,
      objectReference: ctor,
      observedAttributes,
      lifecycleCallbacks,
      disableShadow,
```
So the method get is not working anymore because it check the `ctor` property existence : (https://github.com/jsdom/jsdom/blame/2d82763248296c5f118d6fd890317f5db5de1635/lib/jsdom/living/custom-elements/CustomElementRegistry-impl.js#L217)

```js
  get(name) {
    const definition = this._customElementDefinitions.find(entry => entry.name === name);
    return definition && definition.ctor;
  }
```

### PR DESCRIPTION
**Pull request by fdesforges**
fix #3133 

Test done by creating a npm project using the code from the issue, and checking it 's ok.

I have also run the wpt tests by removing the line in to-run.yaml, the get test are now passing but they are still some others faillure not covered by this PR and this issue, that's why i'm not removing the faillure line in to-run.yaml.

### CODE REVIEW

**Comment by domenic** [comment]
Thanks. Can you ensure we have test coverage so this doesn't happen again, by copying the file into to-upstream but with a "dont-upstream" suffix, and commenting out the ones we can't pass? For example see https://github.com/jsdom/jsdom/blob/master/test/web-platform-tests/to-upstream/html/dom/historical-dont-upstream.html .

**Comment by fdesforges** [comment]
> Thanks. Can you ensure we have test coverage so this doesn't happen again, by copying the file into to-upstream but with a "dont-upstream" suffix, and commenting out the ones we can't pass? For example see https://github.com/jsdom/jsdom/blob/master/test/web-platform-tests/to-upstream/html/dom/historical-dont-upstream.html .

Yeah sure.

File: test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+//             return target[name];
+//         }
+//     });
+//     customElements.define('element-with-attribute-changed-callback', proxy);
+//     assert_array_equals(prototypeCalls, [1, 'connectedCallback', 2, 'disconnectedCallback', 3, 'adoptedCallback', 4, 'attributeChangedCallback']);
+//     assert_array_equals(constructorCalls, [0, 'prototype',
+//                                            5, 'observedAttributes',
+//                                            6, 'disabledFeatures',
+//                                            7, 'formAssociated']);
+// }, 'customElements.define must get "observedAttributes" property on the constructor prototype when "attributeChangedCallback" is present');
```
Comment (ExE-Boss) [inline test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html:340]:
```suggestion
/*
test(function () {
    var constructor = function () {}
    constructor.prototype.attributeChangedCallback = function () { };
    var prototypeCalls = [];
    var callOrder = 0;
    constructor.prototype = new Proxy(constructor.prototype, {
        get: function (target, name) {
            if (name == 'prototype' || name == 'observedAttributes')
                throw 'Unexpected access to observedAttributes';
            prototypeCalls.push(callOrder++);
            prototypeCalls.push(name);
            return target[name];
        }
    });
    var constructorCalls = [];
    var proxy = new Proxy(constructor, {
        get: function (target, name) {
            constructorCalls.push(callOrder++);
            constructorCalls.push(name);
            return target[name];
        }
    });
    customElements.define('element-with-attribute-changed-callback', proxy);
    assert_array_equals(prototypeCalls, [1, 'connectedCallback', 2, 'disconnectedCallback', 3, 'adoptedCallback', 4, 'attributeChangedCallback']);
    assert_array_equals(constructorCalls, [0, 'prototype',
                                           5, 'observedAttributes',
                                           6, 'disabledFeatures',
                                           7, 'formAssociated']);
}, 'customElements.define must get "observedAttributes" property on the constructor prototype when "attributeChangedCallback" is present');
*/
```

File: test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html [hunk 2]
```diff
@@ -0,0 +1,755 @@
...
+//         }
+//     });
+//     assert_throws_js(TypeError,
+//         () => customElements.define('element-with-throwing-callback-3', proxy));
+//     assert_array_equals(calls2, ['connectedCallback', 'disconnectedCallback',
+//                                  'adoptedCallback', 'attributeChangedCallback',
+//                                  'formAssociatedCallback', 'formResetCallback'],
+//         'customElements.define must not get callbacks after one of the get throws');
+// }, 'customElements.define must rethrow an exception thrown while getting ' +
+//    'additional formAssociated callbacks on the constructor prototype');
```
Comment (ExE-Boss) [inline test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html:546]:
```suggestion
/*
test(function () {
    var constructor = function () {}
    var calls = [];
    var err = {name: 'expectedError'};
    var proxy = new Proxy(constructor, {
        get: function (target, name) {
            calls.push(name);
            if (name == 'formAssociated')
                throw err;
            return target[name];
        }
    });
    assert_throws_exactly(err,
        () => customElements.define('element-with-throwing-form-associated', proxy));
    assert_array_equals(calls, ['prototype', 'disabledFeatures', 'formAssociated'],
        'customElements.define must get "prototype", "disabledFeatures", and ' +
        '"formAssociated" on the constructor');
}, 'customElements.define must rethrow an exception thrown while getting ' +
   'formAssociated on the constructor prototype');

test(function () {
    var constructor = function () {}
    var prototypeCalls = [];
    constructor.prototype = new Proxy(constructor.prototype, {
        get: function(target, name) {
            prototypeCalls.push(name)
            return target[name];
        }
    });
    var constructorCalls = [];
    var proxy = new Proxy(constructor, {
        get: function (target, name) {
            constructorCalls.push(name);
            if (name == 'formAssociated')
                return 1;
            return target[name];
        }
    });
    customElements.define('element-with-form-associated-true', proxy);
    assert_array_equals(constructorCalls,
        ['prototype', 'disabledFeatures', 'formAssociated'],
        'customElements.define must get "prototype", "disabledFeatures", and ' +
        '"formAssociated" on the constructor');
    assert_array_equals(
        prototypeCalls,
        ['connectedCallback', 'disconnectedCallback', 'adoptedCallback',
         'attributeChangedCallback', 'formAssociatedCallback',
         'formResetCallback', 'formDisabledCallback',
         'formStateRestoreCallback'],
        'customElements.define must get 8 callbacks on the prototype');
}, 'customElements.define must get four additional callbacks on the prototype' +
   ' if formAssociated is converted to true');

test(function () {
    var constructor = function() {};
    var proxy = new Proxy(constructor, {
        get: function(target, name) {
            if (name == 'formAssociated')
                return {};  // Any object is converted to 'true'.
            return target[name];
        }
    });
    var calls = [];
    var err = {name: 'expectedError'};
    constructor.prototype = new Proxy(constructor.prototype, {
        get: function (target, name) {
            calls.push(name);
            if (name == 'formDisabledCallback')
                throw err;
            return target[name];
        }
    });
    assert_throws_exactly(err,
        () => customElements.define('element-with-throwing-callback-2', proxy));
    assert_array_equals(calls, ['connectedCallback', 'disconnectedCallback',
                                'adoptedCallback', 'attributeChangedCallback',
                                'formAssociatedCallback', 'formResetCallback',
                                'formDisabledCallback'],
        'customElements.define must not get callbacks after one of the get throws');

    var calls2 = [];
    constructor.prototype = new Proxy(constructor.prototype, {
        get: function (target, name) {
            calls2.push(name);
            if (name == 'formResetCallback')
                return 43;  // Can't convert to a Function.
            return target[name];
        }
    });
    assert_throws_js(TypeError,
        () => customElements.define('element-with-throwing-callback-3', proxy));
    assert_array_equals(calls2, ['connectedCallback', 'disconnectedCallback',
                                 'adoptedCallback', 'attributeChangedCallback',
                                 'formAssociatedCallback', 'formResetCallback'],
        'customElements.define must not get callbacks after one of the get throws');
}, 'customElements.define must rethrow an exception thrown while getting ' +
   'additional formAssociated callbacks on the constructor prototype');
*/
```

File: test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html [hunk 3]
```diff
@@ -0,0 +1,755 @@
...
+test(function () {
+    assert_equals(customElements.get('existing-custom-element'), undefined);
+    class ExistingCustomElement extends HTMLElement {};
+    customElements.define('existing-custom-element', ExistingCustomElement);
+    assert_equals(customElements.get('existing-custom-element'), ExistingCustomElement);
+}, 'customElements.get return the constructor of the entry with the given name when there is a matching entry.');
+
+// test(function () {
+//     assert_true(customElements.whenDefined('some-name') instanceof Promise);
+// }, 'customElements.whenDefined must return a promise for a valid custom element name');
```
Comment (ExE-Boss) [inline test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html:628]:
```suggestion
/*
test(function () {
    assert_true(customElements.whenDefined('some-name') instanceof Promise);
}, 'customElements.whenDefined must return a promise for a valid custom element name');
*/
```

File: test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html [hunk 4]
```diff
@@ -0,0 +1,755 @@
...
+// }, 'A promise returned by customElements.whenDefined must be resolved with the defined class');
+
+// promise_test(function () {
+//     var promise = customElements.whenDefined('not-resolved-yet-custom-element').then(function (value) {
+//         assert_equals(value, NotResolvedYetCustomElement, 'The promise returned by "whenDefined" must resolve with the defined class once such class is defined');
+//     });
+//     class NotResolvedYetCustomElement extends HTMLElement {};
+//     customElements.define('not-resolved-yet-custom-element', NotResolvedYetCustomElement);
+//     return promise;
+// }, 'A promise returned by customElements.whenDefined must be resolved with the defined class once such class is defined');
```
Comment (ExE-Boss) [inline test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html:752]:
```suggestion
/*
promise_test(function () {
    class PreexistingCustomElement extends HTMLElement { };
    customElements.define('preexisting-custom-element', PreexistingCustomElement);

    var promise = customElements.whenDefined('preexisting-custom-element');
    promise.then(function (value) { promise.resolved = value; }, function (value) { promise.rejected = value; });

    assert_false('resolved' in promise, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
    assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');

    return Promise.resolve().then(function () {
        assert_true('resolved' in promise, 'The promise returned by "whenDefined" must be resolved when a custom element is defined');
        assert_equals(promise.resolved, PreexistingCustomElement,
            'The promise returned by "whenDefined" must be resolved with the constructor of the element when a custom element is defined');
        assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected when a custom element is defined');
    });
}, 'customElements.whenDefined must return a resolved promise when the registry contains the entry with the given name');

promise_test(function () {
    class AnotherExistingCustomElement extends HTMLElement {};
    customElements.define('another-existing-custom-element', AnotherExistingCustomElement);

    var promise1 = customElements.whenDefined('another-existing-custom-element');
    var promise2 = customElements.whenDefined('another-existing-custom-element');
    promise1.then(function (value) { promise1.resolved = value; }, function (value) { promise1.rejected = value; });
    promise2.then(function (value) { promise2.resolved = value; }, function (value) { promise2.rejected = value; });

    assert_not_equals(promise1, promise2);
    assert_false('resolved' in promise1, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
    assert_false('resolved' in promise2, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
    assert_false('rejected' in promise1, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');
    assert_false('rejected' in promise2, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');

    return Promise.resolve().then(function () {
        assert_true('resolved' in promise1, 'The promise returned by "whenDefined" must be resolved when a custom element is defined');
        assert_equals(promise1.resolved, AnotherExistingCustomElement, 'The promise returned by "whenDefined" must be resolved with the constructor of the element when a custom element is defined');
        assert_false('rejected' in promise1, 'The promise returned by "whenDefined" must not be rejected when a custom element is defined');

        assert_true('resolved' in promise2, 'The promise returned by "whenDefined" must be resolved when a custom element is defined');
        assert_equals(promise2.resolved, AnotherExistingCustomElement, 'The promise returned by "whenDefined" must be resolved with the constructor of the element when a custom element is defined');
        assert_false('rejected' in promise2, 'The promise returned by "whenDefined" must not be rejected when a custom element is defined');
    });
}, 'customElements.whenDefined must return a new resolved promise each time invoked when the registry contains the entry with the given name');

promise_test(function () {
    class ElementDefinedAfterWhenDefined extends HTMLElement { };
    var promise = customElements.whenDefined('element-defined-after-whendefined');
    promise.then(function (value) { promise.resolved = value; }, function (value) { promise.rejected = value; });

    assert_false('resolved' in promise, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
    assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');

    var promiseAfterDefine;
    return Promise.resolve().then(function () {
        assert_false('resolved' in promise, 'The promise returned by "whenDefined" must not be resolved until the element is defined');
        assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected until the element is defined');
        assert_equals(customElements.whenDefined('element-defined-after-whendefined'), promise,
            '"whenDefined" must return the same unresolved promise before the custom element is defined');
        customElements.define('element-defined-after-whendefined', ElementDefinedAfterWhenDefined);
        assert_false('resolved' in promise, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
        assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');

        promiseAfterDefine = customElements.whenDefined('element-defined-after-whendefined');
        promiseAfterDefine.then(function (value) { promiseAfterDefine.resolved = value; }, function (value) { promiseAfterDefine.rejected = value; });
        assert_not_equals(promiseAfterDefine, promise, '"whenDefined" must return a resolved promise once the custom element is defined');
        assert_false('resolved' in promiseAfterDefine, 'The promise returned by "whenDefined" must not be resolved until the end of the next microtask');
        assert_false('rejected' in promiseAfterDefine, 'The promise returned by "whenDefined" must not be rejected until the end of the next microtask');
    }).then(function () {
        assert_true('resolved' in promise, 'The promise returned by "whenDefined" must be resolved when a custom element is defined');
        assert_equals(promise.resolved, ElementDefinedAfterWhenDefined,
            'The promise returned by "whenDefined" must be resolved with the constructor of the element when a custom element is defined');
        assert_false('rejected' in promise, 'The promise returned by "whenDefined" must not be rejected when a custom element is defined');

        assert_true('resolved' in promiseAfterDefine, 'The promise returned by "whenDefined" must be resolved when a custom element is defined');
        assert_equals(promiseAfterDefine.resolved, ElementDefinedAfterWhenDefined,
            'The promise returned by "whenDefined" must be resolved with the constructor of the element when a custom element is defined');
        assert_false('rejected' in promiseAfterDefine, 'The promise returned by "whenDefined" must not be rejected when a custom element is defined');
    });
}, 'A promise returned by customElements.whenDefined must be resolved by "define"');

promise_test(function () {
    class ResolvedCustomElement extends HTMLElement {};
    customElements.define('resolved-custom-element', ResolvedCustomElement);
    return customElements.whenDefined('resolved-custom-element').then(function (value) {
        assert_equals(value, ResolvedCustomElement, 'The promise returned by "whenDefined" must resolve with the defined class');
    });
}, 'A promise returned by customElements.whenDefined must be resolved with the defined class');

promise_test(function () {
    var promise = customElements.whenDefined('not-resolved-yet-custom-element').then(function (value) {
        assert_equals(value, NotResolvedYetCustomElement, 'The promise returned by "whenDefined" must resolve with the defined class once such class is defined');
    });
    class NotResolvedYetCustomElement extends HTMLElement {};
    customElements.define('not-resolved-yet-custom-element', NotResolvedYetCustomElement);
    return promise;
}, 'A promise returned by customElements.whenDefined must be resolved with the defined class once such class is defined');
*/

**Review by ExE-Boss — COMMENTED** [review]
<delete>

**Comment by domenic** [comment]
Please ignore @ExE-Boss's review. Any style of comments is acceptable.

@ExE-Boss, please stop reviewing contributions to this repository, especially from non-maintainers; you are not a maintainer.

**Comment by fdesforges** [comment]
Oops sorry for the wrong naming

**Review by TimothyGu — APPROVED** [review]

File: test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html [hunk 5]
```diff
@@ -0,0 +1,755 @@
+<!DOCTYPE html>
```
Comment (ExE-Boss) [inline test/web-platform-tests/to-upstream/custom-elements/CustomElementRegistry-dont-upstream.html:1]: If you wait for <https://github.com/jsdom/jsdom/pull/3139>, then this file becomes unnecessary.

**Review by ExE-Boss — COMMENTED** [review]
This bug was my fault because I missed this in <https://github.com/jsdom/jsdom/pull/2956>.

**Pull request merged**